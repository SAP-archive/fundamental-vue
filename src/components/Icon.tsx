import { Component, Prop, Base } from '@/core';

const sizeMapping = {
  s: 'Small',
  m: 'Medium',
  l: 'Large',
  xl: 'Extra Large',
};
type IconSize = keyof (typeof sizeMapping);
export const IconSizes = Object.keys(sizeMapping) as IconSize[];

interface Props {
  name?: string;
  size?: IconSize | null;
}

@Component('Icon')
export class Icon extends Base<Props> {
  @Prop('icon name', { type: String, required: true })
  public name!: string;

  @Prop('icon size', { acceptableValues: IconSizes, type: String, default: null })
  public size!: IconSize | null;

  public render() {
    return <span class={this.classes} />;
  }

  private get classes() {
    return {
      [`sap-icon--${this.name}`]: true,
      // Sizes
      'sap-icon--s': this.size === 's',
      'sap-icon--m': this.size === 'm',
      'sap-icon--l': this.size === 'l',
      'sap-icon--xl': this.size === 'xl',
    };
  }
}
