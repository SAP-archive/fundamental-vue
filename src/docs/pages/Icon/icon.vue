<title>Icon usage and size options</title>

<template>
  <div>
    <FdIcon name="cart" size="s" />
    <FdIcon name="cart" />
    <!-- default -->
    <FdIcon name="cart" size="m" />
    <FdIcon name="cart" size="l" />
    <FdIcon name="cart" size="xl" />
  </div>
</template>
