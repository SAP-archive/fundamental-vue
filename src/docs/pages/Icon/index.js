export default {
  relatedComponents: ["fd-icon"]
};
