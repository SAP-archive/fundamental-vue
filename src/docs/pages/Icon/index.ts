import { ExampleCollectionFunction } from '../types';

export const plugin: ExampleCollectionFunction = ({ Icon }) => {
    return { componentStatus: 'stable', icon: 'nutrition-activity', relatedComponents: [Icon] };
};
