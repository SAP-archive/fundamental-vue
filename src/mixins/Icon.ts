import { iconClass, IconName } from '@/lib';
import Vue from 'vue';

export interface IconProps {
  icon?: IconName;
}

export const IconMixin = Vue.extend({
  props: {
    icon: { type: String, default: null },
  },
  computed: {
    iconClassName(): string | null {
      const icon = this.icon;
      if(icon == null) {
        return null;
      }
      return iconClass(icon);
    },
  },
});
