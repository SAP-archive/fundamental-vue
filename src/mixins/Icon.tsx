import {
  Component,
  Prop,
} from 'vue-property-decorator';
import {
  iconClass,
  IconName,
} from '@/lib';
import { Api } from '@/api';
import TsxComponent from '@/vue-tsx';

export interface IconProps {
  icon?: IconName;
}

@Component({
  name: 'Icon',
})
export class Icon extends TsxComponent<IconProps> {
  @Api.Prop('icon name', prop => prop.type(String))
  @Prop({ type: String, default: null })
  public icon!: IconName | null;

  public get iconClassName(): string | null {
    const icon = this.icon;
    if(icon == null) {
      return null;
    }
    return iconClass(icon);
  }
}
