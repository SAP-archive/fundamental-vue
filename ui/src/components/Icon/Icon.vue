<template>
  <span :class="classes" />
</template>

<script lang="ts">
import Vue from "vue";
import { PropValidator } from "vue/types/options";

const sizeMapping = {
  s: "s",
  m: "m",
  l: "l",
  xl: "xl"
};
type IconSize = keyof (typeof sizeMapping);
export const IconSizes = Object.keys(sizeMapping);
const IsValidSize = (value: string) => IconSizes.indexOf(value) >= 0;

export default Vue.extend({
  name: "FdIcon",
  props: {
    name: {
      type: String,
      required: true
    } as PropValidator<string>,
    size: {
      type: String,
      default: null,
      validator: IsValidSize
    } as PropValidator<IconSize | null>
  },
  computed: {
    classes(): object {
      const size =
        this.size == null ? {} : { [`sap-icon--${this.size}`]: true };
      return {
        [`sap-icon--${this.name}`]: true,
        ...size
      };
    }
  }
});
</script>
