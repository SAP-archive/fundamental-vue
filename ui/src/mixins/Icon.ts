import Vue from "vue";
import { PropValidator } from "vue/types/options";
import { iconClass, IconName } from "@/lib";

export interface IconProps {
  icon?: IconName;
}

export default Vue.extend({
  props: {
    icon: {
      type: String,
      default: null
    } as PropValidator<string | null>
  },
  computed: {
    // We are always returning an array of class names.
    // This is done because of the following reasons:
    // - By returning an array we can return an empty array in case icon is null.
    //   This makes it easier for consumers to use this mixin.
    // - We may return more than one class in the futre. Possible use case is to
    //   add another prop that specifies the icon size.
    iconClasses(): string[] {
      return this.icon != null ? [iconClass(this.icon)] : [];
    }
  }
});
