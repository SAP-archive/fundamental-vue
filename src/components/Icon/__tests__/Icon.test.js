import { mount } from "@vue/test-utils";
import Icon from "../Icon.vue";

describe("Icon Test Scripts", () => {
  const name = "cart";
  const size = "s";
  const icon = mount(Icon, {
    propsData: {
      name,
      size
    }
  });

  test("Renders correctly", () => {
    expect(icon.element).toMatchSnapshot();
  });

  test("Render cart icon based on the icon name", () => {
    expect(icon.classes(`sap-icon--${name}`)).toBe(true);
  });

  test("Render icon based on the size mentioned", () => {
    expect(icon.classes(`sap-icon--${size}`)).toBe(true);
  });
});
