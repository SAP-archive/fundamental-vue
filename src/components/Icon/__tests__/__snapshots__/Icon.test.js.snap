// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Test Scripts Renders correctly 1`] = `
<span
  class="sap-icon--cart sap-icon--s"
/>
`;
